Add custom routes docs (rewrites, redirects, headers) (#14887)
vercel/next.js: docs/api-reference/next.config.js/headers.md

---
description: Add custom HTTP headers to your Next.js app.
---

# Headers

Headers allow you to set custom HTTP headers for an incoming request path.

> To use Redirects you can use the `headers` key in `next.config.js`:

Reviewer: "Redirects" seems oddly placed here

> To use Redirects you can use the `headers` key in `next.config.js`:

Reviewer: ```suggestion
To use Headers you can use the `headers` key in `next.config.js`:
```

```js
module.exports = {
  async headers() {
    return [
      {
        source: '/about',
        headers: [
          {
            key: 'x-custom-header',
            value: 'my custom header value',
          },
          {
            key: 'x-another-custom-header',
            value: 'my other custom header value',
          },
        ],
      },
      ,
    ]
  },
}
```

> `headers` is an `async` function that expects an array to be returned holding objects with `source` and `headers`.

Reviewer: Seems odd to codify `async` but not `function`. Let's just drop 'em?
Also add properties suffix.

```suggestion
`headers` is an async function that expects an array to be returned holding objects with `source` and `headers` properties:
```

> - `source` is the incoming request path pattern.

Reviewer: What "pattern"? This should link to our docs on pattern matching.

Author: We don't have docs on pattern matching afaik

> - `headers` is an array of header objects holding `key` and `value`

Reviewer: ```suggestion
- `headers` is an array of header objects with the `key` and `value` properties.
```
